File: task-manager-api/src/models/user_model.py
import bcrypt

class UserModel:
    def __init__(self, db):
        self.db = db

    def find_by_id(self, user_id):
        row = self.db.execute(
            "SELECT id, username, email FROM users WHERE id = ?", (user_id,)
        ).fetchone()
        return dict(row) if row else None

    def find_by_username(self, username):
        row = self.db.execute(
            "SELECT * FROM users WHERE username = ?", (username,)
        ).fetchone()
        return dict(row) if row else None

    def insert(self, username, email, password_hash):
        try:
            cursor = self.db.execute(
                "INSERT INTO users (username, password, email) VALUES (?, ?, ?)",
                (username, email, password_hash)
            )
            self.db.commit()
            return self.find_by_id(cursor.lastrowid)
        except Exception:
            return None

    def verify_password(self, plain_password, hashed):
        try:
            return bcrypt.checkpw(plain_password.encode(), hashed.encode())
        except Exception:
            return False

File: desafios/desafio-3-skill-refactor-arch/ecommerce-api-legacy/src/models/UserModel.js
const bcrypt = require('bcrypt');
const SALT_ROUNDS = 10;

class UserModel {
    constructor(db) { this.db = db; }

    findByEmail(email) {
        return this.db.prepare('SELECT * FROM users WHERE email = ?').get(email) || null;
    }

    findById(id) {
        return this.db.prepare('SELECT id, name, email FROM users WHERE id = ?').get(id) || null;
    }

    async insert(name, email, password) {
        const hash = await bcrypt.hash(password, SALT_ROUNDS);
        try {
            const result = this.db.prepare(
                'INSERT INTO users (name, email, pass) VALUES (?, ?, ?)'
            ).run(name, email, hash);
            return this.findById(result.lastInsertRowid);
        } catch (err) {
            if (err.message.includes('UNIQUE')) throw new Error('Email já cadastrado');
            throw err;
        }
    }

    async verifyPassword(plain, hash) {
        return bcrypt.compare(plain, hash);
    }
}

module.exports = UserModel;
